File: backend/app/models/billing.py
"""
Billing Models - Invoice, Payment, InsuranceClaim - All fields use camelCase
"""
from sqlalchemy import Column, String, Float, DateTime, JSON, Enum, ForeignKey, Boolean
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.database import Base
from app.schemas.enums import PaymentStatus, PaymentMethod, ClaimStatus


class Invoice(Base):
    __tablename__ = "invoices"

    invoiceId = Column("invoice_id", String, primary_key=True, index=True)  # INV-YYYYMMDD-XXX
    orderId = Column("order_id", String, ForeignKey("orders.order_id"), nullable=False, index=True)
    patientId = Column("patient_id", String, ForeignKey("patients.id"), nullable=False, index=True)
    patientName = Column("patient_name", String, nullable=False)

    # Items (JSON array)
    items = Column(JSON, nullable=False)  # Array of {testCode, testName, quantity, unitPrice, totalPrice}

    # Amounts
    subtotal = Column(Float, nullable=False)
    discount = Column(Float, default=0.0)
    tax = Column(Float, default=0.0)
    total = Column(Float, nullable=False)

    # Payment tracking
    paymentStatus = Column("payment_status", Enum(PaymentStatus), nullable=False, default=PaymentStatus.UNPAID)
    amountPaid = Column("amount_paid", Float, default=0.0)
    amountDue = Column("amount_due", Float, nullable=False)

    # Dates
    createdAt = Column("created_at", DateTime(timezone=True), server_default=func.now())
    dueDate = Column("due_date", DateTime(timezone=True), nullable=True)


class Payment(Base):
    __tablename__ = "payments"

    paymentId = Column("payment_id", String, primary_key=True, index=True)  # PAY-YYYYMMDD-XXX
    orderId = Column("order_id", String, ForeignKey("orders.order_id"), nullable=False, index=True)
    invoiceId = Column("invoice_id", String, ForeignKey("invoices.invoice_id"), nullable=True, index=True)

    amount = Column(Float, nullable=False)
    paymentMethod = Column("payment_method", Enum(PaymentMethod), nullable=False)

    paidAt = Column("paid_at", DateTime(timezone=True), nullable=False)
    receivedBy = Column("received_by", String, nullable=False)
    receiptGenerated = Column("receipt_generated", Boolean, default=False)
    notes = Column(String, nullable=True)

    # Relationship for eager loading
    order = relationship("Order", foreign_keys=[orderId])


class InsuranceClaim(Base):
    __tablename__ = "insurance_claims"

    claimId = Column("claim_id", String, primary_key=True, index=True)  # CLM-YYYYMMDD-XXX
    orderId = Column("order_id", String, ForeignKey("orders.order_id"), nullable=False, index=True)
    invoiceId = Column("invoice_id", String, ForeignKey("invoices.invoice_id"), nullable=False, index=True)
    patientId = Column("patient_id", String, ForeignKey("patients.id"), nullable=False, index=True)

    insuranceProvider = Column("insurance_provider", String, nullable=False)
    insuranceNumber = Column("insurance_number", String, nullable=False)

    claimAmount = Column("claim_amount", Float, nullable=False)
    approvedAmount = Column("approved_amount", Float, nullable=True)

    claimStatus = Column("claim_status", Enum(ClaimStatus), nullable=False, default=ClaimStatus.SUBMITTED)

    submittedDate = Column("submitted_date", DateTime(timezone=True), nullable=False)
    processedDate = Column("processed_date", DateTime(timezone=True), nullable=True)

    denialReason = Column("denial_reason", String, nullable=True)
    notes = Column(String, nullable=True)

File: frontend/src/types/billing.ts
/**
 * Billing and Financial Management Types
 */

import { ICONS } from '@/utils';

export type PaymentMethod =
  | 'cash'
  | 'credit-card'
  | 'debit-card'
  | 'insurance'
  | 'bank-transfer'
  | 'mobile-money';

/**
 * Payment method option configuration for UI components
 * Single source of truth for all payment method selections
 */
export interface PaymentMethodOption {
  /** Payment method value */
  value: PaymentMethod;
  /** Display label */
  label: string;
  /** Icon name for visual representation */
  icon: string;
  /** Whether this method is currently enabled */
  enabled: boolean;
}

/**
 * All available payment methods with their configuration
 * Single source of truth - update this array to change payment options across the app
 *
 * To enable/disable a payment method, change the `enabled` flag.
 */
export const PAYMENT_METHOD_OPTIONS: PaymentMethodOption[] = [
  { value: 'cash', label: 'Cash', icon: ICONS.dataFields.cash, enabled: true },
  { value: 'mobile-money', label: 'Mobile Money', icon: ICONS.ui.smartphone, enabled: true },
  { value: 'credit-card', label: 'Credit Card', icon: ICONS.dataFields.creditCard, enabled: false },
  { value: 'debit-card', label: 'Debit Card', icon: ICONS.dataFields.creditCard, enabled: false },
  { value: 'insurance', label: 'Insurance', icon: ICONS.ui.shield, enabled: false },
  { value: 'bank-transfer', label: 'Bank Transfer', icon: ICONS.dataFields.wallet, enabled: false },
];

/**
 * Returns only the enabled payment methods for use in UI selection
 */
export const getEnabledPaymentMethods = (): PaymentMethodOption[] => {
  return PAYMENT_METHOD_OPTIONS.filter(method => method.enabled);
};

/**
 * Default payment method (first enabled method)
 */
export const getDefaultPaymentMethod = (): PaymentMethod => {
  const enabledMethods = getEnabledPaymentMethods();
  return enabledMethods.length > 0 ? enabledMethods[0].value : 'cash';
};

export type ClaimStatus = 'submitted' | 'processing' | 'approved' | 'denied' | 'paid';

export interface InvoiceItem {
  testCode: string;
  testName: string;
  quantity: number;
  unitPrice: number;
  totalPrice: number;
}

export interface Invoice {
  // Identities
  invoiceId: number; // Primary ID, displayed as INV{id}
  id?: number; // Alias for compatibility
  orderId: number;
  patientId: number;
  patientName?: string;

  // Status
  status: 'pending' | 'paid' | 'overdue' | 'cancelled' | 'unpaid';
  paymentStatus?: 'unpaid' | 'paid'; // Deprecated alias

  // Financials
  items: InvoiceItem[];
  subtotal: number;
  discount: number;
  tax: number;
  total: number;
  totalAmount?: number; // Alias
  amountPaid: number;
  paidAmount?: number; // Alias
  amountDue: number;

  // Dates
  createdAt: string;
  dueDate?: string;
}

export interface Payment {
  paymentId: number; // Integer ID, displayed as PAY{id}
  orderId: number;
  invoiceId: number | null;
  amount: number;
  paymentMethod: PaymentMethod;
  paidAt: string;
  receivedBy: string; // Backend returns string user ID
  receiptGenerated: boolean;
  notes?: string;
  // Computed fields from relationships
  orderTotalPrice?: number;
  numberOfTests?: number;
  patientName?: string;
}

export interface InsuranceClaim {
  claimId: number; // Integer ID, displayed as CLM{id}
  orderId: number;
  invoiceId: number;
  patientId: number;
  insuranceProvider: string;
  insuranceNumber: string;
  claimAmount: number;
  approvedAmount?: number;
  claimStatus: ClaimStatus;
  submittedDate: string;
  processedDate?: string;
  denialReason?: string;
  notes?: string;
}
